test/LoginStrategy.test.ts

```diff
@@ -350,6 +350,186 @@ it('Should support old format did for off chain claims', async () => {
+    ensResolver.address,
+    didContract.address,
+    ensRegistry.address,
+    false,
+    ['test']
+  );
+  const token = await iam.claimsService?.createIdentityProof();
+  const payload = {
+    iss: `did:ethr:${userAddress}`,
+    claimData: {
+      blockNumber: 4242,
+    },
+    sub: '',
+  };
+
+  if (!token) {
+    expect(false).toBeTruthy();
+    return;
+  }
+
+  await loginStrategy?.validate(token, payload, (_, user, err) => {
+    expect(err).toEqual('User does not have any roles.');
+  });
+});
+
+it('Should validate user with no roles, if no accepted roles defined', async () => {
+  const { loginStrategy } = preparePassport(
+    provider,
+    ensResolver.address,
+    didContract.address,
+    ensRegistry.address,
+    false
+  );
+  const token = await iam.claimsService?.createIdentityProof();
+  const payload = {
+    iss: `did:ethr:${userAddress}`,
+    claimData: {
+      blockNumber: 4242,
+    },
+    sub: '',
+  };
+
+  if (!token) {
+    expect(false).toBeTruthy();
+    return;
+  }
+
+  await loginStrategy?.validate(token, payload, (_, user, err) => {
+    expect(err).toBe(undefined);
+    expect(user);
+  });
+});
+
+it('Should verify only accepted roles, if includeAllRoles is false', async () => {
+  const claim1: RoleEIP191JWT = {
+    payload: {
+      did: 'did:ethr:volta:0x0000000000000000000000000000000000000001',
+      signer: 'did:ethr:volta:0x0000000000000000000000000000000000000001',
+      claimData: {
+        fields: {},
+        claimType: 'test',
+        claimTypeVersion: 1,
+      },
+      iss: 'did:ethr:volta:0x0000000000000000000000000000000000000001',
+    },
+    eip191Jwt: 'skdjnskdjflksdjlkajsdlkajs',
+  };
+  const claim2: RoleEIP191JWT = {
+    payload: {
+      did: 'did:ethr:volta:0x0000000000000000000000000000000000000001',
+      signer: 'did:ethr:volta:0x0000000000000000000000000000000000000001',
+      claimData: {
+        fields: {},
+        claimType: 'test2',
+        claimTypeVersion: 1,
+      },
+      iss: 'did:ethr:volta:0x0000000000000000000000000000000000000001',
+    },
+    eip191Jwt: 'skdjnskdjflksdjlkajsdlkajs',
+  };
+  const { loginStrategy } = preparePassport(
+    provider,
+    ensResolver.address,
+    didContract.address,
+    ensRegistry.address,
+    false,
+    ['test']
+  );
+  const token = await iam.claimsService?.createIdentityProof();
+  const payload = {
+    iss: `did:ethr:${userAddress}`,
+    claimData: {
+      blockNumber: 4242,
+    },
+    sub: '',
+  };
+
+  if (!token) {
+    expect(false).toBeTruthy();
+    return;
+  }
+
+  jest
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    .spyOn((loginStrategy as any).credentialResolver, 'eip191JwtsOf')
+    .mockResolvedValueOnce([claim1, claim2]);
+
+  const consoleListener = jest.spyOn(console, 'log');
+  await loginStrategy?.validate(token, payload, () => {
+    expect(consoleListener).toBeCalledWith(
+      'includeAllRoles: false, verifying only accepted roles'
+    );
```

Comment:
I think using `consoleListener` is an okay solution, but not great because the log could lie (e.g. it could be in the wrong place). I noticed that the `done` callback param in the `validate` function has a `user` param. https://github.com/energywebfoundation/passport-did-auth/blob/fe0f171aa4b1fcef4cd151819bb3d3d2dd368047/lib/LoginStrategy.ts#L211
Could we assert against that `user` param perhaps? It seems to contain the verified roles result? https://github.com/energywebfoundation/passport-did-auth/blob/fe0f171aa4b1fcef4cd151819bb3d3d2dd368047/lib/LoginStrategy.ts#L295

Reply:
@nichonien what do you mean by "nested scenarios"?

Reply:
I mean scenarios where we need to mock calls that cannot be directly accessed (for ex. If we want to mock getRoleDeifition in IssuerVerification) or are purely primitive types.

Reply:
Yes, I thought that this would be possible:
> So for includeAllRoles to be asserted against user param it should be some actual role claims to be issued and get verified.

But you'd saying that the above is quite challenging, because we can't mock or spy on primitive types. Is this right?

From what I can see, if `ClaimVerifier` was mocked, then the role claim wouldn't have actually be issued & verified as `verifier.getVerifiedRoles()` could be mocked. Does this sound right or not really?

Reply:
> But you'd saying that the above is quite challenging, because we can't mock or spy on primitive types. Is this right?
Yes, mocking is hard. With actual claims, test might take time since it would be validating them through `Volta` test network or may get  in pending state.

`ClaimVerifier` can be mocked if it is instantiated independently (not for LoginStrategy instance as verifier is primitive type) and I was not able to mock `verifier.getVerifiedRoles()`. Also if you mock `verifier.getVerifiedRoles()` and get some expected output aligning to `includeAllRoles` it doesn't seem to be a fit test as it would never allow the validation ever to happen (filtering based on `includeAllRoles` param). Does this answer you concern, or I didn't get it actually?

Reply:
Ah I see, the `ClaimVerifier` instantiation is internal to `LoginStrategy`. Yes, that does make mocking it challenging/unfeasible. Ok, I propose that we leave it as is for now and come back to improve the tests in the future if necessary.